clippy_lints/src/indexing_slicing.rs

```diff
@@ -0,0 +1,201 @@
+//! lint on indexing and slicing operations
+
+use crate::consts::{constant, Constant};
+use crate::utils::higher::Range;
+use crate::utils::{self, higher};
+use rustc::hir::*;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::ast::RangeLimits;
+
+/// **What it does:** Checks for out of bounds array indexing with a constant
+/// index.
+///
+/// **Why is this bad?** This will always panic at runtime.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = [1,2,3,4];
+///
+/// // Bad
+/// x[9];
+/// &x[2..9];
+///
+/// // Good
+/// x[0];
+/// x[3];
+/// ```
+declare_clippy_lint! {
+    pub OUT_OF_BOUNDS_INDEXING,
+    correctness,
+    "out of bounds constant indexing"
+}
+
+/// **What it does:** Checks for usage of indexing or slicing. Does not report
+/// if we can tell that the indexing or slicing operations on an array are in
+/// bounds.
```

Comment:
This is not really true (yet). In the examples below `x[2]` is listed as a bad example, but we can definitely tell, that `2` is in bounds. Maybe rewrite the second sentence as a problem and reformulate  the "What it does"-part so it states, that `get[_mut]()` should be used instead of indexing/slicing with `[a..b]`

Reply:
There should definitely be an example that shows an array so this statement is clearer. It should be true though that this lint does not report _on an array_ if we can tell that the operation is in bounds.